### PR TITLE
Add required vars to config

### PR DESCRIPTION
**Pull request by JanPietrzyk**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
         }
     },
     "require": {
-        "php": ">=7",
+        "php": ">=7.2",
```
Comment (mitelg) [inline composer.json:17]: already PHP8 compatible? maybe use ^ instead?

File: psalm.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0"?>
+<psalm
+    errorLevel="8"
```
Comment (mitelg) [inline psalm.xml:3]: why starting so low? :smile:

File: composer.json (hunk 1 above)
Comment (JanPietrzyk) [reply composer.json:17]: no release, no compatibility. There are still undocumented changes, so I would like to wait. The phar is, btw not limited here

File: psalm.xml (hunk 2 above)
Comment (JanPietrzyk) [reply psalm.xml:3]: Didn't look into it too much. There are still a lot of warnings to get rid of. Some by design - that need thorough changes; some are just busy work. But the codebase passes currently

File: README.md [hunk 3]
```diff
@@ -231,6 +233,18 @@ dev-ops/common/actions/test.sh
...
+
+It may be necessary to require a placeholder to be set, but can't set right away. One such example might be a system dependent path. PSH allows you to communicate this to the user by adding this:
+
+```xml
+<placeholder>
+    <require name="FOO" description="Foo needs to be a reference to bar"/>
+</placeholder>
+```
+
+Now unless foo is set, it is not possible to execute any psh script. The description is optional and can be omitted.
```
Comment (Crease29) [inline README.md:246]:
```suggestion
Although `FOO` is now set, it is not possible to execute any psh script. The description is optional and can be omitted.
```

File: src/Application/Application.php [hunk 4]
```diff
@@ -267,4 +222,94 @@ private function showScriptNotFoundListing(ScriptNotFoundException $ex, array $s
             $this->showListing($scripts);
         }
     }
+
+    private function printHead(Config $config): void
```
Comment (Crease29) [inline src/Application/Application.php:226]: What's the difference between `head` and `header`?

File: src/Application/Application.php [hunk 5]
```diff
@@ -267,4 +222,94 @@ private function showScriptNotFoundListing(ScriptNotFoundException $ex, array $s
...
+                ->createConfig($this->rootDirectory, $inputArgs);
+        } catch (InvalidParameterException $e) {
+            $this->notifyError($e->getMessage() . "\n");
+
+            throw ExitSignal::error();
+        } catch (InvalidArgumentException $e) {
+            $this->notifyError("\n" . $e->getMessage() . "\n");
+
+            throw ExitSignal::error();
+        }
```
Comment (Crease29) [inline src/Application/Application.php:281]:
```suggestion
        } catch (InvalidParameterException | InvalidArgumentException $e) {
            $this->notifyError($e->getMessage() . "\n");

            throw ExitSignal::error();
        }
```

File: src/Application/ExitSignal.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    public const RESULT_SUCCESS = 0;
+
+    public const RESULT_ERROR = 1;
+
+    /**
+     * @var int
+     */
+    private $signal;
+
+    public static function success()
```
Comment (Crease29) [inline src/Application/ExitSignal.php:19]:
```suggestion
    public static function success(): self
```

File: src/Config/Config.php [hunk 7]
```diff
@@ -33,11 +39,13 @@ class Config
      * @param ConfigEnvironment[] $environments
      */
     public function __construct(
+//        EnvironmentResolver $resolver,
```
Comment (Crease29) [inline src/Config/Config.php:42]: Can this be removed?

File: src/Config/EnvironmentResolver.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+use Dotenv\Repository\Adapter\ServerConstAdapter;
+use Symfony\Component\Process\Process;
+use function array_keys;
+use function pathinfo;
+
+class EnvironmentResolver
+{
+    /**
+     * @param DotenvFile[] $dotenvPaths
+     * @return ValueProvider[}
```
Comment (Crease29) [inline src/Config/EnvironmentResolver.php:17]:
```suggestion
     * @return ValueProvider[]
```

File: src/ScriptRuntime/BashCommand.php [hunk 9]
```diff
@@ -12,13 +12,10 @@ class BashCommand implements ProcessCommand
...
     /**
-     * @var string
+     * @var string|null
      */
     private $warning;
 
-    /**
-     * @param string $warning
-     */
     public function __construct(Script $script, string $warning = null)
```
Comment (Crease29) [inline src/ScriptRuntime/BashCommand.php:19]:
```suggestion
    public function __construct(Script $script, ?string $warning = null)
```

**Review by Crease29 — CHANGES_REQUESTED** [review]
For fixing the static method invocation, I can recommend the PHP Inspections (EA Extended) plugin for PhpStorm. You can search for this issue in the whole project and fix it quite easily.
image

File: src/Config/EnvironmentResolver.php (hunk 8 above)
Comment (JanPietrzyk) [reply src/Config/EnvironmentResolver.php:17]: fixed in #109

File: src/Config/Config.php (hunk 7 above)
Comment (JanPietrzyk) [reply src/Config/Config.php:42]: Is already fixed in #109

File: src/Application/ExitSignal.php (hunk 6 above)
Comment (JanPietrzyk) [reply src/Application/ExitSignal.php:19]: Whooops pushed to #109

File: src/Application/Application.php (hunk 5 above)
Comment (JanPietrzyk) [reply src/Application/Application.php:281]: nice. pushed to #109

File: src/Application/Application.php (hunk 4 above)
Comment (JanPietrzyk) [reply src/Application/Application.php:226]: A Problem in naming. The header is a configured flash - like the enterprise in our project. The head is the output of this in combination with config files.... I'll think of something....

File: README.md (hunk 3 above)
Comment (JanPietrzyk) [reply README.md:246]: This reverses the meaning, doesn't it?

**Comment by JanPietrzyk** [comment]
> For fixing the static method invocation, I can recommend the PHP Inspections (EA Extended) plugin for PhpStorm. You can search for this issue in the whole project and fix it quite easily.
> image

I used the cs fixer for that

File: README.md (hunk 3 above)
Comment (Crease29) [reply README.md:246]: Sorry, you're right. I misunderstood the sentence at first.

File: src/Application/Application.php (hunk 4 above)
Comment (JanPietrzyk) [reply src/Application/Application.php:226]: I merged the methods

**Review by Crease29 — APPROVED** [review]

File: psalm.xml (hunk 2 above)
Comment (JanPietrzyk) [reply psalm.xml:3]: done in #109

**Comment by JanPietrzyk** [comment]
Will be merged with #108

**Pull request closed without merging**